File: scratch/deploy_backend.py
import paramiko
import os

def deploy_backend():
    host = "65.20.85.75"
    port = 22
    username = "graceandforce"
    password = "wvpi2!ZnTcV];ncy"

    print("Connecting to Vultr server via SSH...")
    client = paramiko.SSHClient()
    client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
    
    try:
        client.connect(host, port, username, password, timeout=10)
        
        print("Uploading backend files...")
        sftp = client.open_sftp()
        sftp.put("backend/server.js", "/home/graceandforce/debate-engine/backend/server.js")
        sftp.put("backend/package.json", "/home/graceandforce/debate-engine/backend/package.json")
        sftp.put("backend/package-lock.json", "/home/graceandforce/debate-engine/backend/package-lock.json")
        sftp.close()

        print("Restarting PM2 backend...")
        
        script = f"""
export NVM_DIR="$HOME/.nvm"
[ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"

cd /home/graceandforce/debate-engine/backend
npm install
pm2 restart grace-api
"""

        stdin, stdout, stderr = client.exec_command(script)
        
        for line in iter(stdout.readline, ""):
            print(line.encode('utf-8', 'ignore').decode('utf-8'), end="")
            
        for line in iter(stderr.readline, ""):
            print("ERROR: " + line.encode('utf-8', 'ignore').decode('utf-8'), end="")

        exit_status = stdout.channel.recv_exit_status()
        print(f"\\nCommand finished with exit status: {exit_status}")
        
    except Exception as e:
        print(f"Failed to execute: {str(e)}")
    finally:
        client.close()

if __name__ == "__main__":
    deploy_backend()

File: deploy_backend.js
const { Client } = require('ssh2');
const fs = require('fs');
const conn = new Client();

const textFiles = [
  ['backend/server_prod.js', '/home/graceandforce/debate-engine/backend/server_prod.js'],
  ['backend/utils/email.js', '/home/graceandforce/debate-engine/backend/utils/email.js'],
];

const dataFiles = [
  'backend/data/english_questions.json',
  'backend/data/mathematics_questions.json',
  'backend/data/science_questions.json',
  'backend/data/social_science_questions.json',
  'backend/data/ct_ai_questions.json',
];

conn.on('ready', () => {
  conn.exec('mkdir -p /home/graceandforce/debate-engine/backend/data /home/graceandforce/debate-engine/backend/utils', (err, s) => {
    if (err) throw err;
    s.on('close', () => {
      const allFiles = [
        ...textFiles.map(([l, r]) => ({ local: l, remote: r })),
        ...dataFiles.map(f => ({ local: f, remote: '/home/graceandforce/debate-engine/' + f })),
      ];
      let idx = 0;
      const uploadNext = () => {
        if (idx >= allFiles.length) {
          console.log('All files uploaded. Restarting server...');
          conn.exec('source ~/.nvm/nvm.sh && cd /home/graceandforce/debate-engine/backend && pm2 restart grace-api', (e2, s2) => {
            if (e2) throw e2;
            s2.on('data', d => process.stdout.write(d.toString())).on('close', () => conn.end());
          });
          return;
        }
        const { local, remote } = allFiles[idx];
        const content = fs.readFileSync(local, 'utf8');
        conn.sftp((err, sftp) => {
          if (err) throw err;
          const ws = sftp.createWriteStream(remote);
          ws.on('close', () => {
            console.log('Uploaded ' + local);
            idx++;
            sftp.end();
            uploadNext();
          });
          ws.write(content);
          ws.end();
        });
      };
      uploadNext();
    });
  });
}).connect({ host: '65.20.85.75', port: 22, username: 'graceandforce', password: 'wvpi2!ZnTcV];ncy' });
